### PR TITLE
CUMULUS-2394 -- Update granule/pdr writes to not allow old workflows to overwrite new records

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

Addresses [CUMULUS-2394: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2394)

- **CUMULUS-2394**
  - Updated PDR and Granule writes to check the step function start_time against
      the `createdAt` field  for each record to ensure old execution records do not
      overwrite newer executions

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Adhoc testing
- [ ] Integration tests

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,14 +8,28 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
+
+- **CUMULUS-2394**
+  - Updated PDR and Granule writes to check the step function start_time against
+      the `createdAt` fild  for each record to ensure old execution records do not
+      overwrite newer executions
+
 ### Changed
 
-- `<prefix>-lambda-api-gateway` IAM role used by API Gateway Lambda now supports accessing all buckets defined in your `buckets` variable except "internal" buckets
+- `<prefix>-lambda-api-gateway` IAM role used by API Gateway Lambda now supports
```
Comment (Jkovarik) [inline CHANGELOG.md:25]: Updates below are using re-wrap on the CHANGELOG

File: packages/errors/src/index.ts [hunk 2]
```diff
@@ -113,41 +113,36 @@ export const MissingBucketMap = createErrorType('MissingBucketMap');
...
 
-export const InvalidChecksum = createErrorType('InvalidChecksum');
+export const DeletePublishedGranule = createErrorType('DeletePublishedGranule');
 
 export const DuplicateFile = createErrorType('DuplicateFile');
 
-export const UnexpectedFileSize = createErrorType('UnexpectedFileSize');
+export const EcsStartTaskError = createErrorType('EcsStartTaskError');
 
-/**
```
Comment (Jkovarik) [inline packages/errors/src/index.ts:132]: These comments don't appear to match the error class.    Removing and re-alpha-ing the uncommented errors.

File: packages/api/models/pdrs.js [hunk 3]
```diff
@@ -5,6 +5,7 @@ const get = require('lodash/get');
 const log = require('@cumulus/common/log');
 const { getCollectionIdFromMessage } = require('@cumulus/message/Collections');
 const { getMessageExecutionArn } = require('@cumulus/message/Executions');
+const { getCreatedAt } = require('@cumulus/message/Message');
```
Comment (Jkovarik) [inline packages/api/models/pdrs.js:8]: This should be re-implemented like the feature-rds branch, this helper is already in message/src/workflows.ts.

Comment (Jkovarik) [reply packages/api/models/pdrs.js:8]: Update from feature branch: 55dd9454f

File: packages/message/src/workflows.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { CumulusMessageError } from '@cumulus/errors';
```
Comment (Jkovarik) [inline packages/message/src/workflows.ts:1]: This is implemented in the feature branch, porting back to get needed types.

File: packages/api/models/pdrs.js [hunk 5]
```diff
@@ -118,20 +119,23 @@ class Pdr extends Manager {
...
-      } catch (error) {
-        if (error.name && error.name.includes('ConditionalCheckFailedException')) {
-          const executionArn = getMessageExecutionArn(cumulusMessage);
-          log.info(`Did not process delayed 'running' event for PDR: ${pdrRecord.pdrName} (execution: ${executionArn})`);
-          return undefined;
-        }
-        throw error;
+      updateParams.ConditionExpression += ' and (execution <> :execution OR progress < :progress)';
+    }
+    let result;
```
Comment (markdboyd) [inline packages/api/models/pdrs.js:127]: I'm not sure I understand the merit of the refactor here? Having the `result` variable vs not having it?

Comment (Jkovarik) [reply packages/api/models/pdrs.js:127]: Hmm, there isn't a good reason, good catch. 

25894a56c

File: packages/api/tests/models/granules/test-granule-reporting.js [hunk 6]
```diff
@@ -218,6 +220,82 @@ test('_validateAndStoreGranuleRecord() will allow a running status to replace a
...
+  };
+
+  await granuleModel._validateAndStoreGranuleRecord(updatedGranule);
+
+  const fetchedItem = await granuleModel.get({ granuleId: granule.granuleId });
+
+  t.is(fetchedItem.status, 'completed');
+});
+
+test('_validateAndStoreGranuleRecord() will allow an newer final status to replace a final status for an older execution ', async (t) => {
```
Comment (markdboyd) [inline packages/api/tests/models/granules/test-granule-reporting.js:275]:
```suggestion
test('_validateAndStoreGranuleRecord() will allow a newer final status to replace a final status for an older execution ', async (t) => {
```

File: packages/api/tests/models/granules/test-granule-reporting.js [hunk 7]
```diff
@@ -218,6 +220,82 @@ test('_validateAndStoreGranuleRecord() will allow a running status to replace a
...
+  };
+
+  await granuleModel._validateAndStoreGranuleRecord(updatedGranule);
+
+  const fetchedItem = await granuleModel.get({ granuleId: granule.granuleId });
+
+  t.is(fetchedItem.status, 'running');
+});
+
+test('_validateAndStoreGranuleRecord() will not allow an older final status to replace a final status for a newer execution ', async (t) => {
```
Comment (markdboyd) [inline packages/api/tests/models/granules/test-granule-reporting.js:249]:
```suggestion
test('_validateAndStoreGranuleRecord() will not allow a final status for an older execution to replace a final status for a newer execution ', async (t) => {
```

File: packages/api/tests/models/granules/test-granule-reporting.js [hunk 8]
```diff
@@ -218,6 +220,82 @@ test('_validateAndStoreGranuleRecord() will allow a running status to replace a
   t.is(fetchedItem.status, 'running');
 });
 
+test('_validateAndStoreGranuleRecord() will not allow an older final status to replace a running status for a newer execution ', async (t) => {
```
Comment (markdboyd) [inline packages/api/tests/models/granules/test-granule-reporting.js:223]:
```suggestion
test('_validateAndStoreGranuleRecord() will not allow a final status for an older execution to replace a running status for a newer execution ', async (t) => {
```

File: packages/api/tests/models/test-pdrs-model.js [hunk 9]
```diff
@@ -250,6 +250,46 @@ test(
...
+      name: pdrName,
+    };
+
+    await pdrsModel.storePdrFromCumulusMessage(initialMsg);
+
+    const newMsg = createPdrMessage({
+      execution: randomId('newExec'),
+      stateMachine,
+      numRunningExecutions: 2,
+      status: 'running',
```
Comment (markdboyd) [inline packages/api/tests/models/test-pdrs-model.js:278]: I think we should pass in the same `Date.now()` as `createdAtTime` to this factory invocation to be sure that it is an older time

File: packages/api/tests/models/test-pdrs-model.js [hunk 10]
```diff
@@ -250,6 +250,46 @@ test(
   }
 );
 
+test(
+  'storePdrFromCumulusMessage does not update if PDR record is from an older, prior execution',
```
Comment (markdboyd) [inline packages/api/tests/models/test-pdrs-model.js:254]:
```suggestion
  'storePdrFromCumulusMessage does not update PDR record if update is from an older execution',
```

File: packages/api/tests/models/test-pdrs-model.js [hunk 11]
```diff
@@ -292,6 +332,46 @@ test(
   }
 );
 
+test(
+  'storePdrFromCumulusMessage does not update if PDR record is from an older, prior completed execution',
```
Comment (markdboyd) [inline packages/api/tests/models/test-pdrs-model.js:336]:
```suggestion
  'storePdrFromCumulusMessage does not update PDR record if update is from an older completed execution',
```

File: packages/api/tests/models/test-pdrs-model.js [hunk 12]
```diff
@@ -292,6 +332,46 @@ test(
...
+      name: pdrName,
+    };
+
+    await pdrsModel.storePdrFromCumulusMessage(initialMsg);
+
+    const newMsg = createPdrMessage({
+      execution: randomId('newExec'),
+      stateMachine,
+      numRunningExecutions: 2,
+      status: 'failed',
```
Comment (markdboyd) [inline packages/api/tests/models/test-pdrs-model.js:360]: Same here, I think we should pass in `createdAtTime` to be sure that this message test fixture is older than the one previously created in this test

File: packages/errors/src/index.ts (hunk 2 above)
Comment (markdboyd) [reply packages/errors/src/index.ts:132]: I don't follow. Did you just alphabetize the contents of this file?

File: packages/message/tests/test-workflows.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
+const test = require('ava');
+
+const { getMessageWorkflowStartTime } = require('../workflows');
+
+test('getCreatedAt returns the workflow start time', (t) => {
```
Comment (markdboyd) [inline packages/message/tests/test-workflows.js:7]:
```suggestion
test('getMessageWorkflowStartTime
 returns the workflow start time', (t) => {
```

File: packages/message/tests/test-workflows.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+  const testMessage = {
+    cumulus_meta: {
+      workflow_start_time: 123456,
+    },
+  };
+  const result = getMessageWorkflowStartTime(testMessage);
+  t.is(result, 123456);
+});
+
+test('getMessageGranules throws when a start time is not present in the message', (t) => {
```
Comment (markdboyd) [inline packages/message/tests/test-workflows.js:17]:
```suggestion
test('getMessageWorkflowStartTime throws when a start time is not present in the message', (t) => {
```

**Review by markdboyd — COMMENTED** [review]
Several very small comments to address, but otherwise looks great!

File: packages/errors/src/index.ts (hunk 2 above)
Comment (Jkovarik) [reply packages/errors/src/index.ts:132]: Yes, and removed the block comment for this and for `DeletePublishedGranule` as they make no sense.

File: packages/api/tests/models/test-pdrs-model.js (hunk 12 above)
Comment (Jkovarik) [reply packages/api/tests/models/test-pdrs-model.js:361]: 8ecae9657

File: packages/api/tests/models/test-pdrs-model.js (hunk 9 above)
Comment (Jkovarik) [reply packages/api/tests/models/test-pdrs-model.js:278]: Fair point, this is reliant on the helper default for logic that's being explicitly testing.

f251723ef

File: CHANGELOG.md [hunk 15]
```diff
@@ -22,14 +22,30 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
+  `https://custom-cmr-host.com`).
 - Added `buckets` variable to `tf-modules/archive`
 - **CUMULUS-2328**
   - Added `tf-modules/s3_credentials` module which contains resources to attach the `/s3-credentials` endpoint to an API gateway
 
+### Fixed
+
+- **CUMULUS-2394**
+  - Updated PDR and Granule writes to check the step function start_time against
+      the `createdAt` fild  for each record to ensure old execution records do not
```
Comment (markdboyd) [inline CHANGELOG.md:36]:
```suggestion
      the `createdAt` field  for each record to ensure old execution records do not
```

File: CHANGELOG.md [hunk 16]
```diff
@@ -22,14 +22,30 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
+  be used to set to an arbitray  host for making CMR requests (e.g.
+  `https://custom-cmr-host.com`).
 - Added `buckets` variable to `tf-modules/archive`
 - **CUMULUS-2328**
   - Added `tf-modules/s3_credentials` module which contains resources to attach the `/s3-credentials` endpoint to an API gateway
 
+### Fixed
+
+- **CUMULUS-2394**
+  - Updated PDR and Granule writes to check the step function start_time against
```
Comment (markdboyd) [inline CHANGELOG.md:35]:
```suggestion
  - Updated PDR and Granule writes to check the step function `workflow_start_time` against
```

**Comment by Jkovarik** [comment]
@markdboyd  Thanks for the feedback!   I believe I've addressed your concerns, take a look at let me know what you think.

File: CHANGELOG.md [hunk 17]
```diff
@@ -22,14 +22,30 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
 - Added `buckets` variable to `tf-modules/archive`
 - **CUMULUS-2328**
   - Added `tf-modules/s3_credentials` module which contains resources to attach the `/s3-credentials` endpoint to an API gateway
 
+### Fixed
+
+- **CUMULUS-2394**
+  - Updated PDR and Granule writes to check the step function `workflow_start_time` against
+      the `createdAt` field  for each record to ensure old execution records do not
+      overwrite newer executions
```
Comment (markdboyd) [inline CHANGELOG.md:37]:
```suggestion
      the `createdAt` field  for each record to ensure old records do not
      overwrite newer ones
```

Nit to make it clearer that we're not talking about execution records

Comment (Jkovarik) [reply CHANGELOG.md:37]: 👍🏻

**Review by markdboyd — APPROVED** [review]

**Pull request merged**